pkg/proxy/nftables/proxier.go

```diff
@@ -0,0 +1,1633 @@
+}
+
+func isServiceChainName(chainString string) bool {
+	// The chains returned from servicePortChainNameBase and
+	// servicePortEndpointChainNameBase will always have at least one "/" in them.
+	// Since none of our "stock" chain names use slashes, we can distinguish them this
+	// way.
+	return strings.Contains(chainString, "/")
+}
+
+func isAffinitySetName(set string) bool {
+	return strings.HasPrefix(set, servicePortEndpointAffinityNamePrefix)
+}
+
+// This is where all of the nftables calls happen.
+// This assumes proxier.mu is NOT held
+func (proxier *Proxier) syncProxyRules() {
+	proxier.mu.Lock()
+	defer proxier.mu.Unlock()
+
+	// don't sync rules till we've received services and endpoints
+	if !proxier.isInitialized() {
+		klog.V(2).InfoS("Not syncing nftables until Services and Endpoints have been received from master")
+		return
+	}
+
+	//
+	// Below this point we will not return until we try to write the nftables rules.
+	//
+
+	// Keep track of how long syncs take.
+	start := time.Now()
+	defer func() {
+		metrics.SyncProxyRulesLatency.Observe(metrics.SinceInSeconds(start))
+		klog.V(2).InfoS("SyncProxyRules complete", "elapsed", time.Since(start))
+	}()
+
+	serviceUpdateResult := proxier.svcPortMap.Update(proxier.serviceChanges)
+	endpointUpdateResult := proxier.endpointsMap.Update(proxier.endpointsChanges)
+
+	klog.V(2).InfoS("Syncing nftables rules")
+
+	success := false
+	defer func() {
+		if !success {
+			klog.InfoS("Sync failed", "retryingTime", proxier.syncPeriod)
+			proxier.syncRunner.RetryAfter(proxier.syncPeriod)
+		}
+	}()
+
+	// If there are sufficiently-stale chains left over from previous transactions,
+	// try to delete them now.
+	if len(proxier.staleChains) > 0 {
+		oneSecondAgo := start.Add(-time.Second)
+		tx := proxier.nftables.NewTransaction()
+		deleted := 0
+		for chain, modtime := range proxier.staleChains {
+			if modtime.Before(oneSecondAgo) {
+				tx.Delete(&knftables.Chain{
+					Name: chain,
+				})
+				delete(proxier.staleChains, chain)
+				deleted++
+			}
+		}
+		if deleted > 0 {
+			klog.InfoS("Deleting stale nftables chains", "numChains", deleted)
+			err := proxier.nftables.Run(context.TODO(), tx)
+			if err != nil {
+				// We already deleted the entries from staleChains, but if
+				// the chains still exist, they'll just get added back
+				// (with a later timestamp) at the end of the sync.
+				klog.ErrorS(err, "Unable to delete stale chains; will retry later")
+				// FIXME: metric
+			}
+		}
+	}
+
+	// Now start the actual syncing transaction
+	tx := proxier.nftables.NewTransaction()
+	proxier.setupNFTables(tx)
+
+	// We need to use, eg, "ip daddr" for IPv4 but "ip6 daddr" for IPv6
+	ipX := "ip"
+	ipvX_addr := "ipv4_addr" //nolint:stylecheck // var name intentionally resembles value
+	if proxier.ipFamily == v1.IPv6Protocol {
+		ipX = "ip6"
+		ipvX_addr = "ipv6_addr"
+	}
+
+	// We currently fully-rebuild our sets and maps on each resync
+	tx.Flush(&knftables.Set{
```

Comment:
Instead of having `kubeFirewallSet` of type string we can directly define it of type `knftables.Set`, right ?
Then we can use the same variable in every sync rather than creating the structure again to reduce the memory allocations  and deallocations ?

```go
var kubeFirewallSet = &knftables.Set{
    Name: "firewall",
}
```

I see `Set` field of type `knftables.Element` is string type, maybe that should also be of type `&knftables.Set` instead of string (better consistency) ?

Likewise for all the `sets` and `maps`, maybe ?

Just thinking out loud here 🙂